File: engines/AnapsidFiles/FedraSourceSelection.py
from utils import *
import collections
import sys
import os
import parseQuery
from SPARQLWrapper import SPARQLWrapper, JSON
from SPARQLWrapper.Wrapper import QueryResult
import rdflib
from rdflib.plugins.sparql import prepareQuery
from rdflib.term import URIRef, Literal, Variable
import parseEndpoints, parseQuery
from services import Argument, Triple
import random

def obtainInstance(candidateSources, elements, collections):
        for t in candidateSources:
            fs = candidateSources[t]
            i = 0
            for f in fs:
                elements.add((t, i)) 
                for e in f:
                    if not (e in collections):
                        ts = set()
                    else:
                        ts = collections[e]
                    ts.add((t, i))
                    collections[e] = ts
                i = i + 1

def getMinimalSetCovering(elements, collections, r):
        elementsToCover = { x for x in elements }
        selected = set()
        while len(elementsToCover) > 0:
            c = []
            for k in collections:
                if (len(c) == 0): 
                    c.append(k)
                else:
                    tmp1 = { x for x in elementsToCover if x in collections[c[0]] }
                    tmp2 = { x for x in elementsToCover if x in collections[k] }
                    if (len(tmp1) < len(tmp2)):
                        c = [ k ]
                    elif len(tmp1) == len(tmp2):
                        c.append(k)
            if r:
                e = random.choice(c)
            else: 
                e = c[0]
            elementsToCover = { x for x in elementsToCover if not(x in collections[e])}
            selected.add(e)
            del collections[e]
        return selected

def select(candidateSources, selected):
        ss = {}
        for t in candidateSources:
            fs = candidateSources[t]
            es = set()
            for f in fs:
                nf = set(f)
                nf.intersection_update(selected)
                es.add(frozenset(nf))
            ss[t] = es
        return ss

def filter(candidateSources, bgp):
    result = {}
    for t in candidateSources:
        if t in bgp:
            result[t] = candidateSources[t]
    return result

def update(candidateSources, selected):
    for t in candidateSources:
        if t in selected:
            candidateSources[t] = selected[t]


def getTriples(query):
        a = query.algebra
        ts = getTriplesRec(a)
        return ts

def getTriplesRec(a):
        #print str(a.name)
        #print str(type(a))
        if (a.name == 'Union') or (a.name == 'Join') or (a.name == 'LeftJoin'):
            return getTriplesRec(a.p1) + getTriplesRec(a.p2)
        elif (a.name == 'Minus'):
            return getTriplesRec(a.p1)
        elif (a.name == 'BGP'):
            return a.triples
        elif (a.name == 'ConstructQuery') or (a.name == 'SelectQuery') or (a.name == 'Filter') or (a.name == 'Project') or (a.name == 'Group') or (a.name == 'OrderBy') or (a.name == 'ToMultiSet') or (a.name == 'Graph'):
            return getTriplesRec(a.p)
        else:
            return []

class TriplePatternFragment:
    def __init__(self, t, ds):
        self.triple = t
        self.dataset = ds
        self.endpoints = []

    def addSource(self, e):
        self.endpoints.append(e)

    def getDataset(self):
        return self.dataset

    def contains(self, other):
        c = isinstance(other, TriplePatternFragment)
        #print 'other is TPF? '+str(c)
        c = c and (self.dataset == other.dataset)
        #print 'same dataset? '+str(c)
        c = c and self.weaker(self.triple.predicate, other.triple.predicate)
        c = c and self.weaker(self.triple.subject, other.triple.subject)
        c = c and self.weaker(self.triple.theobject, other.triple.theobject)
        return c

    def containedBy(self, other):
        c = isinstance(other, TriplePatternFragment)
        c = c and (self.dataset == other.dataset)
        c = c and self.weaker(other.triple.predicate, self.triple.predicate)
        c = c and self.weaker(other.triple.subject, self.triple.subject)
        c = c and self.weaker(other.triple.theobject, self.triple.theobject)
        return c

    def canAnswer(self, triple):                                               
        c = isinstance(triple, Triple)
        c = c and self.compatible(triple.predicate, self.triple.predicate)
        c = c and self.compatible(triple.subject, self.triple.subject)
        c = c and self.compatible(triple.theobject, self.triple.theobject)
        return c

    def containsTriple(self, triple):
        c = isinstance(triple, Triple)
        c = c and self.weaker(self.triple.predicate, triple.predicate)
        c = c and self.weaker(self.triple.subject, triple.subject)
        c = c and self.weaker(self.triple.theobject, triple.theobject)
        return c

    def exactMatch(self, triple): 
        m = isinstance(triple, Triple)
        m = m and self.exactM(triple.predicate, self.triple.predicate)           
        m = m and self.exactM(triple.subject, self.triple.subject)               
        m = m and self.exactM(triple.theobject, self.triple.theobject)
        return m

    def weaker(self, t1, t2):
        #print 't1 constant: '+str(t1.constant)+' t1 name: '+str(t1.name)
        #print 't2 constant: '+str(t2.constant)+' t2 name: '+str(t2.name)
        b = (not t1.constant or (t2.constant and (t1.name == t2.name)))
        #print 't1 weaker than t2: '+str(b)
        return b

    def compatible(self, t1, t2):
        b = not (t1.constant and t2.constant) or (t1.name == t2.name)
        return b

    def exactM(self, t1, t2):                                                     
        b = (not t1.constant and not t2.constant) or (t1.constant and t2.constant and (t1.name == t2.name))
        return b

    def getAllSources(self):
        return self.endpoints

    def __eq__(self, other):
        e = not (other == None) and isinstance(other, TriplePatternFragment)
        e = e and self.dataset == other.dataset
        e = e and self.exactM(self.triple.predicate, other.triple.predicate)
        e = e and self.exactM(self.triple.subject, other.triple.subject)
        e = e and self.exactM(self.triple.theobject, other.triple.theobject)
        return e

    def __hash__(self):
        predicate = None
        subject = None
        object = None
        if (self.triple.predicate.constant):
            predicate = self.triple.predicate
        if (self.triple.subject.constant):
            subject = self.triple.subject
        if (self.triple.theobject.constant):
            object = self.triple.theobject
        return (hash((self.dataset, predicate, subject, object)))

    def __repr__(self):
        return '<'+str(self.triple)+', '+str(self.dataset)+', '+str(self.endpoints)+'>'

class FedraSourceSelection:

    def __init__(self, tl, el, props, ps):
        self.tripleList = []
        for t in tl:
            ns = Argument(getUri(t.subject, ps), t.subject.constant)
            np = Argument(getUri(t.predicate, ps), t.predicate.constant)
            no = Argument(getUri(t.theobject, ps), t.theobject.constant)
            self.tripleList.append(Triple(ns, np, no))
        self.endpointsList = el
        self.selectedSources = {}
        self.loadFragmentDefinitions(props['FragmentsDefinitionFolder'], props['FragmentsSources'])
        self.loadEndpoints(props['EndpointsFile'])
        self.random = (props['Random'] == 'true')
        #print 'end of the constructor'
        #print self.fragments

    def getSelectedSources(self):
        return self.selectedSources

    def getEndpoints(self, selectedFragments):
        fs = set()
        #print 'self.endpointsList: '+str(self.endpointsList)
        for fragmentList in selectedFragments:
            f = set()
            for fragment in fragmentList:
                l = fragment.getAllSources()
                #print 'l: '+str(l)
                #print '[s[0] for s in self.endpointsList]: '+str([s[0] for s in self.endpointsList])
                f = [s for s in l if '<'+s+'>' in { f for (f, s) in self.endpointsList}]
            #print 'f: '+str(f)
            fs.add(frozenset(f))
        return fs

    def sourceSelectionPerTriple(self):
        #print 1
        candidates = {}
        for t in self.tripleList:
            #print 'triple: '+str(t)
            selectedFragments = set()
            for fn in self.fragments:
                f = self.fragments[fn]
                #print 'considering fragment: '+str(fn)
                #print 'f: '+str(f)
                if f.canAnswer(t):
                    #print 'it can answer'
                    redundantFragments = set()
                    toAdd = True
                    includeWith = []
                    for l in selectedFragments:
                        lAux = list(l)
                        f2 = lAux[0]
                        #print 'already included fragment: '+str(f2.triple)
                        #print 'f.containsTriple(t): '+str(f.containsTriple(t))
                        #print 'f2.containsTriple(t): '+str(f2.containsTriple(t))
                        #print 'f.contains(f2): '+str(f.contains(f2))
                        #print 'f.containedBy(f2): '+str(f.containedBy(f2))
                        if (f.containsTriple(t) and f2.containsTriple(t) and (f.contains(f2) or f.containedBy(f2))):
                            includeWith.append(l)
                            toAdd = False
                        elif f.containedBy(f2):
                            toAdd = False
                            break
                        elif f.contains(f2):
                            redundantFragments.add(l)
                    #print 'redundantFragments: '+str(redundantFragments)
                    #print 'toAdd: '+str(toAdd)
                    for l in redundantFragments:
                        selectedFragments.remove(l)
                    for fl in includeWith:
                        if not (f in fl):
                            selectedFragments.remove(fl)
                            s = set(fl)
                            s.add(f)
                            selectedFragments.add(frozenset(s))
                    if toAdd:
                        l = frozenset({f})
                        selectedFragments.add(l)

            endpoints = self.getEndpoints(selectedFragments)
            candidates[t] = endpoints
        return candidates 

    def performSourceSelection(self):
        candidateSources = self.sourceSelectionPerTriple()
        #print 'candidate sources: '+str(candidateSources)
        bgp2 = []
        for t in self.tripleList:
            if len(candidateSources[t]) > 1:
                fragments = candidateSources[t]
                for fragment in fragments:
                    break
                intersection = { f for f in fragment }
                for f in fragments:
                    intersection.intersection_update(f)
                if len(intersection) > 0:
                    fragments = { frozenset(intersection) }
                    candidateSources[t] = fragments
        for t in self.tripleList:
            if len(candidateSources[t]) == 1:
                bgp2.append(t)
        bgpCandidateSources = filter(candidateSources, bgp2)
        elements = set()
        collections = {}
        obtainInstance(candidateSources, elements, collections)
        selectedSubsets = getMinimalSetCovering(elements, collections, self.random)
        bgpCandidateSources = select(bgpCandidateSources, selectedSubsets)
        update(candidateSources, bgpCandidateSources)

        self.selectedSources.clear()
        for t in candidateSources:
            fs = candidateSources[t]
            es = set()
            for f in fs:
                e = random.choice(list(f))
                es.add('<'+str(e)+'>')
            self.selectedSources[t] = es

    def loadEndpoints(self, fileName):
        with open (fileName, 'r') as f:
            for line in f:
                line = line.strip()
                ws = line.split()
                if (len(ws) > 0):
                    fragment = ws[0]
                    i = 1
                    f = self.fragments[fragment]
                    while i < len(ws):
                        endpoint = ws[i]
                        f.addSource(endpoint)
                        i = i + 1

    def loadFragmentDefinitions(self, folder, fileName):
        datasets = {}
        self.fragments = {}
        with open (fileName, 'r') as f:
            for line in f:
                line = line.strip()
                ws = line.split()
                if (len(ws) > 0):
                    fragment = ws[0]
                    ds = ws[1]
                    datasets[fragment] = ds

        content = os.listdir(folder)
        self.viewsDefinition = {}
        for g in content:
            path = folder+'/'+g
            f = open(path)
            viewStr = f.read()
            f.close()
            view = prepareQuery(viewStr)
            t = getTriple(view)
            i = path.rfind('/') + 1
            j = path.rfind('.')
            j = len(path) if j < 0 else j
            name = path[i:j]
            ds = datasets[name]
            self.fragments[name] = TriplePatternFragment(t, ds)

def getTriple(query):
    ts = getTriples(query)
    t = ts[0]
    (s, p, o) = t
    s = Argument(s.n3(), not isinstance(s, Variable))
    p = Argument(p.n3(), not isinstance(p, Variable))
    o = Argument(o.n3(), not isinstance(o, Variable))
    t = Triple(s,p,o)
    return t

#def main(argv):

#if __name__ == '__main__':
#    main(sys.argv)

File: engines/fedX/src/com/fluidops/fedx/optimizer/FedraSourceSelection.java
package com.fluidops.fedx.optimizer;

import java.util.*;
import java.io.*;

import org.openrdf.query.QueryLanguage;
//import org.openrdf.query.Query;
import org.openrdf.repository.sparql.SPARQLRepository;
import org.openrdf.repository.RepositoryConnection;
import org.openrdf.query.parser.sparql.SPARQLParserFactory;
import org.openrdf.query.parser.QueryParser;
import org.openrdf.query.parser.ParsedQuery;
import org.openrdf.query.algebra.TupleExpr;
import org.openrdf.query.algebra.StatementPattern;
import org.openrdf.query.algebra.Var;
import org.openrdf.queryrender.sparql.SparqlTupleExprRenderer;
import org.openrdf.query.BooleanQuery;
import org.openrdf.query.impl.EmptyBindingSet;
import com.fluidops.fedx.structures.Endpoint;
import com.fluidops.fedx.Config;
import com.fluidops.fedx.evaluation.TripleSource;
import org.openrdf.query.algebra.helpers.StatementPatternCollector;
/*
import com.hp.hpl.jena.graph.Triple;
import com.hp.hpl.jena.graph.Node;
import com.hp.hpl.jena.sparql.core.BasicPattern;
import com.hp.hpl.jena.query.QueryFactory;
import com.hp.hpl.jena.query.Query;
import com.hp.hpl.jena.sparql.algebra.*;
import com.hp.hpl.jena.sparql.algebra.op.*;
import com.hp.hpl.jena.sparql.syntax.*;
import com.hp.hpl.jena.query.QueryExecution;
import com.hp.hpl.jena.query.QueryExecutionFactory;
import com.hp.hpl.jena.sparql.engine.http.QueryEngineHTTP;
import com.hp.hpl.jena.sparql.expr.ExprList;

import com.hp.hpl.jena.util.FileManager;
import com.hp.hpl.jena.util.Locator;*/

class FedraSourceSelection {

    private TupleExpr query;
    private List<StatementPattern> stmts;
    private HashMap<String,Endpoint> endpoints;
    private HashMap<StatementPattern, Set<Endpoint>> selectedSources;
    private ArrayList<ArrayList<StatementPattern>> bgps;
    private TreeMap<String, TriplePatternFragment> fragments;
    private boolean random;

    public FedraSourceSelection(String queryStr, List<StatementPattern> stmts, List<Endpoint> endpoints) {
        QueryParser qp = (new SPARQLParserFactory()).getParser();
        String baseURI = null;
        try {
            ParsedQuery pq = qp.parseQuery(queryStr, baseURI);
            this.query = pq.getTupleExpr();
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
        this.stmts = stmts;
        this.endpoints = new HashMap<String,Endpoint>();
        for (Endpoint e : endpoints) {
            this.endpoints.put(e.getEndpoint(), e);
        }
        //System.out.println("endpoints: "+endpoints);
        this.selectedSources = new HashMap<StatementPattern, Set<Endpoint>>();
        this.bgps = getBGPs(query);
        loadFragments(Config.getConfig().getProperty("FragmentsDefinitionFolder"), Config.getConfig().getProperty("FragmentsSources"));
        loadEndpoints(Config.getConfig().getProperty("EndpointsFile"));
        this.random = Boolean.parseBoolean(Config.getConfig().getProperty("Random"));
        //System.out.println("endpoints: "+endpoints);
        //System.out.println("fragments: "+fragments);
        //System.out.println("endpoints: "+endpoints);
    }

    private static ArrayList<ArrayList<StatementPattern>> getBGPs(TupleExpr query) {

        ArrayList<ArrayList<StatementPattern>> bgps = null;
        try {
            BGPVisitor bgpv = new BGPVisitor();
            query.visit(bgpv);
            bgps =  bgpv.getBGPs();
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
        return bgps;
    }

    public HashMap<StatementPattern, Set<Endpoint>> getSelectedSources() {

        return this.selectedSources;
    }

    private HashSet<TreeSet<Endpoint>> getEndpoints(HashSet<List<TriplePatternFragment>> selectedFragments) {

        //System.out.println("selected fragments: "+selectedFragments);
        //System.out.println("endpoints: "+this.endpoints);
        HashSet<TreeSet<Endpoint>> fs = new HashSet<TreeSet<Endpoint>>();
        for (List<TriplePatternFragment> fragmentList : selectedFragments) {
                TreeSet<Endpoint> f = new TreeSet<Endpoint>(new EndpointComparator());
                for (TriplePatternFragment fragment : fragmentList) {
                    List<String> l = fragment.getAllSources();
                    for (String en : l) {
                        if (this.endpoints.containsKey(en)) {
                            f.add(this.endpoints.get(en));
                        }
                    }
                }
                fs.add(f);
        }
        //System.out.println("selected endpoints: "+fs);
	return fs;
    }

    public HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> sourceSelectionPerTriple() {

        HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidates = new HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>>();
        for (StatementPattern sp : this.stmts) {
            HashSet<List<TriplePatternFragment>> selectedFragments = new HashSet<List<TriplePatternFragment>>();
            //System.out.println("considering sp: "+sp);
            for (String fn : this.fragments.keySet()) {
                TriplePatternFragment f = this.fragments.get(fn);
                if (f.canAnswer(sp)) {
                    HashSet<List<TriplePatternFragment>> redundantFragments = new HashSet<List<TriplePatternFragment>>();
                    boolean toAdd = true;
                    List<List<TriplePatternFragment>> includeWith = new ArrayList<List<TriplePatternFragment>>(); 
                    //System.out.println("trying to add: "+fn+" from dataset: "+f.getDataset());
                    //System.out.println("is f an exact match of sp? "+f.exactMatch(sp));
                    //System.out.println("so far there are "+selectedFragments.size()+ " fragments");
                    for (List<TriplePatternFragment> l :  selectedFragments) {
                        TriplePatternFragment f2 = l.get(0);
                        //System.out.println("checking already added fragment: "+f2.triple+" from dataset: "+f2.getDataset()+" in fragment list: "+l);
                        //System.out.println("is f2 an exact match of sp? "+f2.exactMatch(sp));
                        //System.out.println("f.getDataset().equals(f2.getDataset()): "+f.getDataset().equals(f2.getDataset()));
                        //System.out.println("f.containedBy(f2): "+f.containedBy(f2));
                        //System.out.println("f.contains(f2): "+f.contains(f2));
                        //System.out.println("f.contains(sp): "+f.contains(sp));
                        //System.out.println("f2.contains(sp): "+f2.contains(sp));
                        if (f.contains(sp) && f2.contains(sp) && (f.contains(f2) || f.containedBy(f2))) {
                            includeWith.add(l);
                            toAdd = false;
                        } else if (f.containedBy(f2)) {
                            toAdd = false;                                                                                                                                          
                            break;
                        } else if (f.contains(f2)) {
                            redundantFragments.add(l);
                        }
                    }
                    //System.out.println("removing "+redundantFragments.size()+" fragments");
                    for (List<TriplePatternFragment> l : redundantFragments) {
                        selectedFragments.remove(l);
                    }
                    for (List<TriplePatternFragment> fl : includeWith) {
                        if (!fl.contains(f)) {
                            selectedFragments.remove(fl);
                            fl.add(f);
                            selectedFragments.add(fl);
                        }
                    }
                    if (toAdd) {
                        ArrayList<TriplePatternFragment> l = new ArrayList<TriplePatternFragment>();
                        l.add(f);
                        selectedFragments.add(l);
                    }
                }
            }
            //System.out.println("selected fragments: "+selectedFragments);
            HashSet<TreeSet<Endpoint>> endpoints = getEndpoints(selectedFragments);
            candidates.put(sp, endpoints);
        }
        return candidates;
    }

    private static void obtainInstance(HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidateSources, 
                                       TreeSet<Pair<StatementPattern, Integer>> elements, 
                                       TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>> collections) { 
        // Obtaining the instance of the minimal set covering problem
        for (StatementPattern sp : candidateSources.keySet()) {
            HashSet<TreeSet<Endpoint>> fs = candidateSources.get(sp);
            int i = 0;
            for (TreeSet<Endpoint> f : fs) {
                Pair<StatementPattern, Integer> p = new Pair<StatementPattern, Integer>(sp, i);
                elements.add(p);
                for (Endpoint e : f) {
                    TreeSet<Pair<StatementPattern, Integer>> sps = collections.get(e);
                    if (sps == null) {
                        sps = new TreeSet<Pair<StatementPattern, Integer>>(new PairComparator());
                    }
                    sps.add(p);
                    collections.put(e, sps);
                }
                i++;
            }
        }
    }

    private static TreeSet<Endpoint> getMinimalSetCovering(TreeSet<Pair<StatementPattern, Integer>> elements, 
                                       TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>> collections, boolean random) {
        TreeSet<Pair<StatementPattern, Integer>> elementsToCover = new TreeSet<Pair<StatementPattern, Integer>> (new PairComparator());
        elementsToCover.addAll(elements);
        TreeSet<Endpoint> selected = new TreeSet<Endpoint>(new EndpointComparator());
        //System.out.println("elements: "+elements);
        while(elementsToCover.size() > 0) {
            TreeSet<Endpoint> c = new TreeSet<Endpoint>(new EndpointComparator());
            //System.out.println("collections: "+collections);
            for (Endpoint k : collections.keySet()) {
                if (c.size() == 0) {
                    c.add(k);
                } else {
                    TreeSet<Pair<StatementPattern, Integer>> tmp1 = new TreeSet<Pair<StatementPattern, Integer>>(elementsToCover);
                    tmp1.retainAll(collections.get(c.first()));
                    TreeSet<Pair<StatementPattern, Integer>> tmp2 = new TreeSet<Pair<StatementPattern, Integer>>(elementsToCover);
                    tmp2.retainAll(collections.get(k));
                    if (tmp1.size() < tmp2.size()) {
                        c.clear();
                        c.add(k);
                    } else if (tmp1.size() == tmp2.size()) {
                        c.add(k);
                    }
                }
            }
            Iterator<Endpoint> it = c.iterator();
            Endpoint e = it.next();
            if (random) {
                int elm = (int) (Math.random()*c.size());
                for (int i = 0; i < elm && it.hasNext(); i++) {
                    e = it.next();
                }
            }
            elementsToCover.removeAll(collections.get(e));
            selected.add(e);
            collections.remove(e);
        }
        return selected;
    }

    private static HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> select(HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidateSources, TreeSet<Endpoint> selected) {

        HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> ss = new HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>>();
        for (StatementPattern sp : candidateSources.keySet()) {
            HashSet<TreeSet<Endpoint>> fs = candidateSources.get(sp);
            //System.out.println("selected sources for "+sp+" are: "+fs);
            HashSet<TreeSet<Endpoint>> es = new HashSet<TreeSet<Endpoint>>();
            for (TreeSet<Endpoint> f : fs) {
                f.retainAll(selected);
                es.add(f);
            }
            ss.put(sp, es);
        }
        return ss;
    }

    private static HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> filter(HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidateSources, ArrayList<StatementPattern> bgp) {
        HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> result = new HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> ();
        for (StatementPattern sp : candidateSources.keySet()) {
            if (bgp.contains(sp)) {
                 result.put(sp, candidateSources.get(sp));
            }
        }
        return result;
    }

    private static void update(HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidateSources, HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> selected) {

        for (StatementPattern sp : candidateSources.keySet()) {                                                                                                  
            if (selected.containsKey(sp)) {
                candidateSources.put(sp, selected.get(sp));
            }
        }
    }

    public void performSourceSelection() { 
        
        HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> candidateSources = sourceSelectionPerTriple();
        //System.out.println("candidate sources: "+candidateSources);

        // Priority is given to evaluate triple patterns that belong to the same basic graph pattern in as less endpoints as possible
        for (ArrayList<StatementPattern> bgp : bgps) {
            ArrayList<StatementPattern> bgp2 = new ArrayList<StatementPattern>();
            for (StatementPattern sp : bgp) {
                if (candidateSources.get(sp).size()>1) {
                    HashSet<TreeSet<Endpoint>> fragments = candidateSources.get(sp);
                    Iterator<TreeSet<Endpoint>> it = fragments.iterator();
                    TreeSet<Endpoint> intersection = new TreeSet<Endpoint>(it.next());
                    while (it.hasNext()) {
                        TreeSet<Endpoint> f = it.next();
                        intersection.retainAll(f);
                    }
                    if (intersection.size() > 0) {
                        fragments.clear();
                        fragments.add(intersection);
                        candidateSources.put(sp, fragments);
                    }
                }
            }
            for (StatementPattern sp : bgp) {                                                                                                                                       
                if (candidateSources.get(sp).size()==1) {
                    bgp2.add(sp);
                }
            }
            HashMap<StatementPattern, HashSet<TreeSet<Endpoint>>> bgpCandidateSources = filter(candidateSources, bgp2);
            TreeSet<Pair<StatementPattern, Integer>> elements = new TreeSet<Pair<StatementPattern, Integer>>(new PairComparator());
            TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>> collections = new TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>>(new EndpointComparator());
            obtainInstance(bgpCandidateSources, elements, collections);
            TreeSet<Endpoint> selectedSubsets = getMinimalSetCovering(elements, collections, this.random);
            bgpCandidateSources = select(bgpCandidateSources, selectedSubsets);
            update(candidateSources, bgpCandidateSources);
        }
        //System.out.println("candidate sourcesB: "+candidateSources);

        // Then a global choice is done for all the triple patterns, selecting as few endpoints as possible.
//        TreeSet<Pair<StatementPattern, Integer>> elements = new TreeSet<Pair<StatementPattern, Integer>>(new PairComparator());
//        TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>> collections = new TreeMap<Endpoint, TreeSet<Pair<StatementPattern, Integer>>>(new EndpointComparator());
//        obtainInstance(candidateSources, elements, collections);                                                                                                                
//        TreeSet<Endpoint> selectedSubsets = getMinimalSetCovering(elements, collections, this.random);
//        candidateSources = select(candidateSources, selectedSubsets);
        //System.out.println("candidate sourcesC: "+candidateSources);

        // The remaining choices are done randomly to select one source for each different data fragment
        selectedSources.clear();
        for (StatementPattern sp : candidateSources.keySet()) { 
            HashSet<TreeSet<Endpoint>> fs = candidateSources.get(sp);
            //System.out.println("selected sources for "+sp+" are: "+fs);
            TreeSet<Endpoint> es = new TreeSet<Endpoint>(new EndpointComparator());
            for (TreeSet<Endpoint> f : fs) {
                Iterator<Endpoint> it = f.iterator();
                Endpoint e = it.next();
                if (this.random) {
                    int elm = (int) (Math.random()*f.size());
                    for (int i = 0; i < elm && it.hasNext(); i++) {
                        e = it.next();
                    }
                }
                es.add(e);
            }
            selectedSources.put(sp, es);
        }
        //System.out.println("selected sources: "+selectedSources);
    }

    public void loadEndpoints (String file) {

        try {
            BufferedReader br = new BufferedReader(new FileReader(file));
            String l = br.readLine();
            while (l!=null) {
                StringTokenizer st = new StringTokenizer(l);
                if (st.hasMoreTokens()) {
                    String fragment = st.nextToken();
                    //endpoints.add(endpoint);
                    TriplePatternFragment f = this.fragments.get(fragment);
                    while (st.hasMoreTokens()) {
                        String endpoint = st.nextToken();
                        f.addSource(endpoint);
                    }
                    // this.fragments.put(fragment, f);
                }
                l = br.readLine();
            }
            br.close();
            
        } catch (IOException e) {
            System.err.println("Problems reading file: "+file);
            System.exit(1);
        }
    }

    public static HashMap<String, String> loadSources (String file) {                                                                                                                                       

        HashMap map = new HashMap<String, String>();
        try {
            BufferedReader br = new BufferedReader(new FileReader(file));
            String l = br.readLine();
            while (l!=null) {
                StringTokenizer st = new StringTokenizer(l);
                if (st.hasMoreTokens()) {
                    String fragment = st.nextToken();
                    String source =  st.nextToken();
                    map.put(fragment, source);
                }
                l = br.readLine();
            }
            br.close();

        } catch (IOException e) {
            System.err.println("Problems reading file: "+file);
            System.exit(1);
        }
        return map;
    }


    private static String getQuery(String file) {

        String query = "";
        try {
            BufferedReader br = new BufferedReader(new FileReader(file));
            String l = br.readLine();
            while (l!=null) {
                query = query + l + "\n";
                l = br.readLine();
            }
            br.close();
        } catch (IOException e) {
            System.err.println("Problems reading file: "+file);
            System.exit(1);
        }
        return query;
    }

    private StatementPattern getStatementPattern(TupleExpr te) {

        List<StatementPattern> stmts = StatementPatternCollector.process(te);
        StatementPattern sp = stmts.get(0);
        return sp;
    }

    public void loadFragments(String folder, String file) {

        HashMap<String, String> sources = loadSources(file);
        File f = new File(folder);
        File[] content = f.listFiles();
        this.fragments = new TreeMap<String, TriplePatternFragment>();
        //System.out.println("folder: "+folder);
        QueryParser qp = (new SPARQLParserFactory()).getParser();
        if (content != null) {
            for (File g : content) {
                try {
                    String path = g.getAbsolutePath();
                    //System.out.println("path: "+path);
                    String query = getQuery(path);
                    String baseURI = null;
                    ParsedQuery pq = qp.parseQuery(query, baseURI);
                    TupleExpr te = pq.getTupleExpr();
                    StatementPattern sp = getStatementPattern(te);
                    int i = path.lastIndexOf("/") + 1;
                    int j = path.lastIndexOf(".");
                    j = j < 0 ? path.length() : j;
                    String name = path.substring(i, j);
                    this.fragments.put(name, new TriplePatternFragment(sp, sources.get(name)));
                }  catch (Exception e) {
                    e.printStackTrace();
                    System.exit(1);
                }
            }
        }
    }
} 
